two_factor_otp_auth/controllers/main.py

```diff
@@ -0,0 +1,84 @@
+# Copyright 2019 VentorTech OU
+# License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl-3.0).
+
+import logging
+
+from odoo import http, _
+from odoo.addons.web.controllers.main import Home
+from odoo.http import request
+
+_logger = logging.getLogger(__name__)
+
+try:
+    from ..exceptions import MissingOtpError, InvalidOtpError
+
+except IOError as error:
+    _logger.debug(error)
+
+
+class Login2fa(Home):
+
+    @http.route()
+    def web_login(self, redirect=None, **kw):
+        """
+        Overload core method to start Second Factor validation step
+        """
+        try:
+            response = super(Login2fa, self).web_login(redirect, **kw)
+        except MissingOtpError:
+            # user will coming here while login process is not full successful
+            # come here after success validation of other credentials
+            # to start Second Factor (OTP) validation step
+            response = self._get_response()
+        except InvalidOtpError:
+            message = _("Your security code is wrong")
+            response = self._get_response(message)
+        else:
+            params = request.params
+            if params.get("login_success"):
+                user = request.env.user
+                if user and user.enable_2fa and not user.qr_image_2fa:
+                    # If credentials are Okay, but a user doesn't have
+                    # QR code, that mean it's first success login with
+                    # one-time-password. Now QR Code with it's Secret
+                    # Code can be saved into the user.
+                    values = {
+                        "qr_image_2fa": params.get("qr_code_2fa"),
+                        "secret_code_2fa": params.get("secret_code_2fa"),
+                    }
+                    user.sudo().write(values)
+
+        return response
+
+    @staticmethod
+    def _get_response(message=None):
+        """
+        Method for getting response object which depending on user and values
```

Comment:
Correct grammar mistakes: Method to get response object that depends on user and values

Reply:
done